semgrep/semgrep/join_rule.py

```diff
@@ -0,0 +1,404 @@
+) -> None:
+    for result in semgrep_results:
+        metavars = result.get("extra", {}).get("metavars", {})
+        renamed_metavars = {
+            refs_lookup[result.get("check_id", "")].renames.get(
+                metavar, metavar
+            ): contents
+            for metavar, contents in metavars.items()
+        }
+        result["extra"]["metavars"] = renamed_metavars
+
+
+def create_model_map(
+    semgrep_results: List[Dict[str, Any]]
+) -> Dict[str, Type[BaseModel]]:
+    collections: Dict[str, List[Dict]] = group(
+        semgrep_results, key=lambda item: item.get("check_id")
+    )
+    model_map: Dict[str, Type[BaseModel]] = {}
+    for name, findings in collections.items():
+        metavars = set()
+        for finding in findings:
+            metavars.update(finding.get("extra", {}).get("metavars", {}).keys())
+        model_fields = ["path"] + list(metavars)
+        model_class = model_factory(camel_case(name), model_fields)
+        model_map[name] = model_class
+    return model_map
+
+
+def populate_data(
+    semgrep_results: List[Dict[str, Any]], model_map: Dict[str, Type[BaseModel]]
+) -> None:
+    collections = group(semgrep_results, key=lambda item: item.get("check_id"))
+    for name, findings in collections.items():
+        for finding in findings:
+            model_map[name].create(
+                path=finding.get("path"),
+                raw=json.dumps(finding),
+                **{
+                    metavar: content.get("abstract_content").strip().strip("\"'")
+                    for metavar, content in finding.get("extra", {})
+                    .get("metavars", {})
+                    .items()
+                },
+            )
+
+
+def main(
+    join_rule: Dict[str, Any],
+    targets: List[Path],
+) -> Tuple[List[RuleMatch], List[SemgrepError]]:
+    join_contents = join_rule.get("join", {})
+    semgrep_config_strings = [ref.get("rule") for ref in join_contents.get("refs", [])]
+    config_map = create_config_map(semgrep_config_strings)
+
+    join_rule_refs: List[Ref] = [
+        Ref(
+            id=config_map[ref.get("rule")].id,
+            renames={
+                rename.get("from"): rename.get("to")
+                for rename in ref.get("renames", [])
+            },
+            alias=ref.get("as"),
+        )
+        for ref in join_contents.get("refs", [])
+    ]
+    refs_lookup = {ref.id: ref for ref in join_rule_refs}
+    alias_lookup = {ref.alias: ref.id for ref in join_rule_refs}
+
+    try:
+        conditions = [
+            Condition.parse(condition_string)
+            for condition_string in join_contents.get("on", [])
+        ]
+    except InvalidConditionError as e:
+        return [], [e]
+
+    # Run Semgrep
+    with tempfile.NamedTemporaryFile() as rule_path:
+        yaml.dump({"rules": [rule.raw for rule in config_map.values()]}, rule_path)
+        rule_path.flush()
+        output = semgrep.semgrep_main.invoke_semgrep(
```

Comment:
In the future we can probably be smarter and run the rules that need to be run along with all the other rules and just pass results to join but this is fine for experimental feature :D

Reply:
Yah, it was easier to make a different code path using the unofficial API 😅